qlib/Logger.qm

```diff
@@ -0,0 +1,2733 @@
+            @param count number of files in rotation chain, if count is <=0 then no ratation is performed
+            @param archive pattern to evaluate archive filename; see below for more information
+            @param encoding the encoding of the output file
+
+            The following table explains the characters used in @ref LoggerPattern for both filename and archive parameter
+
+            | !Conversion character | !Meaning |
+            | s | Used to output the OS directory separator, i.e. "\" or "/" |
+            | c | Used to output the appender name |
+            | C | Used to output the appender class name |
+            | E | Used to output the environment variable with name given by option, e.g. \c "%E{HOME}" |
+            | % | The literal percent sign. \c "%%" will print a \c "%" sign |
+
+            The following table explains the characters used in @ref LoggerPattern for archive parameter only
+
+            | !Conversion character | !Meaning |
+            | P | Used to output the current logging path with filename |
+            | p | Used to output the current logging path |
+            | f | Used to output the current logging filename |
+            | i | Used to output the current file index from 1 to count. option may specify @ref sprintf() format, e.g. \c "%.3d" |
+            | n | Used to output the rotation count. Option may specify @ref sprintf() format, e.g. \c "%.3d" |
+
+            For example, the archive pattern \c "%p%f.%i" rotates log file \c "/var/run/log/MyLogger.log" to:
+            \c "/var/run/log/MyLogger.log.1" \c "/var/run/log/MyLogger.log.2" \c "/var/run/log/MyLogger.log.3".
+        */
+        constructor(string name, LoggerLayout layout, string filename, int count, string archive = DEFAULT_ARCHIVE_PATTERN, *string encoding): LoggerAppenderFile(name, layout, "", encoding), LoggerPattern(filename) {
+            m_filename = format(NOTHING);  # no dynamic vars as timestamp
+            m_count = count;
+            setPattern(archive);
+        }
+
+        #! Returns a string for a format field for a pattern-based filename
+        /**
+            @param data input context reference data
+            @param key the format character
+            @param option the format option value, if any
+
+            @return the formatted string
+        */
+        private *string resolveField(auto data, string key, *string option) {
+            *string res;
+            switch (key) {
+                case "c":
+                    res = get_class_name(self);
+                    break;
+                case "C":
+                    res = getName();
+                    break;
+                case "E":
+                    res = ENV{option};
+                    break;
+                case "s":
+                    res = DirSep;
+                    break;
+            }
+            if (data) {  # archive pattern
+                switch 
```

Comment:
this could be much faster with `map` as:
```
# post event to all appenders
map $1.post(event), m_appenders;
```

Because `map` is much faster than `foreach`